[TechDocs] Use an event-stream for the sync endpoint to give access to the logs of the sync process

dhenneke opened the pull request:
We are working on the UX/DX of the "basic" architecture of TechDocs (see also https://github.com/backstage/backstage/pull/6263). Our goal is to have the option to run the docs generation in Backstage instead of in the CI. We did some work to create a trusted execution environment for the generation, but we want the user to be able to access the logs of the process. Which is the scope of this PR 😀.

This PR includes multiple changes. Each is part of individual commits, so it might be easier to review them individually.

* **@backstage/techdocs-common**:
  * Add an optional `logStream` to `GeneratorBase#run(...)` that is logged to and that is forwarded to the `ContainerRunner` interface. This is similar to what the scaffolder does.
* **@backstage/techdocs-backend**:
  * Refactor the `/sync/:namespace/:kind/:name` to return an event-stream and to avoid a long-polling approach.
  * Use the event-stream to transmit the logs that are emitted by the `GeneratorBase`.
* **@backstage/techdocs**:
  * Refactor the client to interact with the event-stream.
  * Provide access to the received log messages in the UI.
  * Update the "initial load" view.
  * Add a "Refresh" button to the page to hot-reload new content.

Some remarks:

* ~~I replaced the old endpoint but I could also do the implementation in a new endpoint to not break the API.~~ -> see https://github.com/backstage/backstage/pull/6341#issuecomment-874915863
* I used the [`eventsource`](https://www.npmjs.com/package/eventsource) package instead of the raw browser API as it supports `headers` and I found it easier to mock in tests. So it should (hopefully) also work for users that use an auth-middleware (/cc https://github.com/backstage/backstage/issues/5041#issuecomment-873891617).
* I hope the change is not too large. I can split it up if required.

### Some Demos

Do an initial load:

https://user-images.githubusercontent.com/720821/124475075-9a1d0b80-dda1-11eb-9de5-5bc7b4f3ea37.mov

Hot-reload the updated content:

https://user-images.githubusercontent.com/720821/124475962-9c339a00-dda2-11eb-9a85-f73947713772.mov

View logs of a failed synchronization:

https://user-images.githubusercontent.com/720821/124475252-c9cc1380-dda1-11eb-9429-b0c998e1c77f.mov


<!-- Please describe what you added, and add a screenshot if possible.
     That makes it easier to understand the change so we can :shipit: faster. -->

#### :heavy_check_mark: Checklist

<!--- Please include the following in your Pull Request when applicable: -->

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

dhenneke commented on packages/techdocs-common/src/stages/generate/index.ts:
```diff
@@ -15,4 +15,8 @@
+export type {
+  GeneratorBuilder,
+  GeneratorBase,
+  GeneratorRunOptions,
```
It wasn't exported before but I find it helpful if it is part of the `api-report.md`.

ottosichert commented:
Hello again @dhenneke 👋 looking through your PR and have one first question:

> Some remarks:
> - I replaced the old endpoint but I could also do the implementation in a new endpoint to not break the API.

With this breaking change it might be possible to run into a situation where `@backstage/plugin-techdocs` and `@backstage/plugin-techdocs-backend` run on different versions, and hence would break.

Would it be possible to find a backwards-compatible way for the same functionality? Either through a new endpoint you mentioned or some kind of fallback?

dhenneke commented:
> Would it be possible to find a backwards-compatible way for the same functionality? Either through a new endpoint you mentioned or some kind of fallback?

The EventSource sends a `accept: text/event-stream` header which we could use to determine what to return. I think I could also make it work such that both options use the same code so we don't run into implementations drifting apart.

dhenneke commented:
As an alternative, I can refactor the new endpoint from a `GET` to a `POST` and keep the old one as a deprecated endpoint. Using a `GET` for an operation with side effects isn't very good after all. Both in terms of semantics and because of different CORS handling.

dhenneke commented:
I re-added the old long-polling behavior if the called doesn't request an event-stream in the latest commit.

ottosichert commented on plugins/techdocs/src/reader/components/useReaderState.ts:
```diff
@@ -216,11 +221,18 @@ export function useReaderState(
-): { state: ContentStateTypes; content?: string; errorMessage?: string } {
+): {
+  state: ContentStateTypes;
+  contentReload: () => void;
```
The `contentReload` strikes back 😄 (https://github.com/backstage/backstage/pull/5975#discussion_r651879164). While we still don't produce side effect in the reducer, we now have two different places (`useReaderState` and `<Reader>`) in which we manually trigger a change by circumventing the state machine.

I'm suggesting to replace the `useAsyncRetry` with an `useEffect` that only fires when the `BUILD_READY_RELOAD` state is being reached. Once the `useEffect` has finished fetching, it should dispatch `BUILD_READY` (or similar). That means the interface to produce side effects is by dispatching states, but the hook itself takes care of return the expected data to the consumer.

dhenneke replied:
I think the main issue comes from the fact that I tried to combine two independent processes in the same hook:
1. Download the most recent content of a page regardless of it being fresh or not.
2. Check if the contents in the e.g. S3 bucket are fresh and if not, automatically refresh it, regardless of whether the content was downloaded or not.

With **2.** having the special desire to automatically redo **1.** if-and-only-if there was nothing to show before. And now the user also wants to be able to redo **1.** whenever he likes. But he doesn't want to infer with the sync process of **2.** while doing it.

I also noticed that the combination of the states sometimes behaves a bit weird and that it might be better to decouple the header (is there a sync and what is their outcome) from the content (is it loading and can it be displayed) by putting **1.** and **2.** into individual hooks and move this special desire of **2.** up into the `<Reader/>` component.

dhenneke replied:
But this change (adding the refresh button when ready) is also not connected to the event-stream functionality. It might be better to just leave it out for now and revisit the case in a different PR.

ottosichert commented on packages/techdocs-common/src/stages/generate/techdocs.ts:
```diff
@@ -136,7 +153,7 @@ export class TechdocsGenerator implements GeneratorBase {
       this.logger.debug(
         `Failed to generate docs from ${inputDir} into ${outputDir}`,
       );
-      this.logger.error(`Build failed with error: ${log}`);
+      childLogger.error(`Build failed with error: ${log}`);
```
The frontend build logs do not contain the actual error that is being thrown below. The preferred option would be to have any errors thrown in this scope also being added to the child logger. (If that's not possible, we should at least log the error in all cases before `throw`ing and exiting the scope)

dhenneke replied:
I also noticed that there are more logs that are not yet captured. For example these `Step 1 of 3: Preparing docs for entity ...` logs or the exception that e.g. docker is not available. The question is whether we want to pipe all these to the event stream and what kind of log messages we want to have in the backend logs. One option could be to skip logging anything to the default logger until we get an exception. Only then, print all previously captured logs. The scaffolder has these persisted in the database as an alternative, but I'm not sure if we want to go that far for TechDocs.

ottosichert reviewed (commented):
Thanks again for the great UX improvement contribution! Only point that I think should at least be addressed it logging the full logger in the frontend, as otherwise the feature would only have limited usability as we would require users to go back to the terminal to find the actual error.

It can easily be reproduced by stopping Docker and attempting to build docs

dhenneke commented on plugins/techdocs-backend/src/service/router.ts:
```diff
@@ -108,109 +117,31 @@ export async function createRouter({
+      () => {
+        if (req.header('accept') !== 'text/event-stream') {
+          console.warn(
+            "The /sync/:namespace/:kind/:name endpoint but the call wasn't done by an EventSource. This behavior is deprecated and will be removed soon. Make sure to update the @backstage/plugin-techdocs package in the frontend to the latest version.",
```
Good question 🤔

dhenneke commented on plugins/techdocs/src/client.ts:
```diff
@@ -192,44 +193,59 @@ export class TechDocsStorageClient implements TechDocsStorageApi {
+
+    return new Promise((resolve, reject) => {
+      const source = new EventSource(url, {
+        withCredentials: true,
```
There are some users who use this [authenticating middleware](https://github.com/backstage/backstage/blob/master/contrib/docs/tutorials/authenticate-api-requests.md) and there have been [repeating discussions](https://github.com/backstage/backstage/issues/5041) about it. The scaffolder for example [uses this](https://github.com/backstage/backstage/blob/cecfe5c7f5a65bba79520146008833176930ba92/plugins/scaffolder/src/api.ts#L211) because one can't add headers to the original `EventSource`. We don't use it and thus I don't know what they really need, but I also don't want to break it for those users. Maybe @benjdlambert can give some input on what to do in this situation.

dhenneke commented:
I refactored it to catch all log messages that occur during a sync. The backend logs only contain an error if something went wrong.

I also updated the case "the initial build went wrong" and now display the error alert so the user can get access to the logs:

![localhost_3000_docs_default_Component_documented-component](https://user-images.githubusercontent.com/720821/125078685-0a3dd100-e0c3-11eb-850e-3d02b502df60.png)

dhenneke commented:
I resolved all merge conflicts. Should be ready for another review.

iamEAP commented on plugins/techdocs-backend/src/service/router.ts:
```diff
@@ -136,114 +142,58 @@ export async function createRouter(
+    let responseHandler: DocsSynchronizerSyncOpts;
+    if (req.header('accept') !== 'text/event-stream') {
+      console.warn(
+        "The call to /sync/:namespace/:kind/:name wasn't done by an EventSource. This behavior is deprecated and will be removed soon. Make sure to update the @backstage/plugin-techdocs package in the frontend to the latest version.",
```
Nice.

OrkoHunter reviewed (approved):
❤️ 

@iamEAP @ottosichert @backstage/techdocs-core over to you for final merge!

iamEAP reviewed (approved):
Yeah, indeed. Thanks @dhenneke! Excited to see this go in.
